### DISCUSSION TITLE
Feature request: "Perceptual Image Compression" / pick optimal compression level

### DISCUSSION THREAD

**Discussion by rayshan** [libvips/libvips#1999] [Ideas]
Hello, thanks for your work on libvips. Would you please consider implementing such a feature? Here are some references I found:

Work at Flickr:
http://code.flickr.net/2015/09/25/perceptual-image-compression-at-flickr/

JPEGmini's papers:
http://spie.org/Publications/Proceedings/Paper/10.1117/12.872231
http://spie.org/Publications/Proceedings/Paper/10.1117/12.872230

Implementation based on JPEGmini's papers:
https://github.com/dwbuiten/smallfry

**Comment by jcupitt** [comment]
I agree, it seems interesting. The smallfry code looks easy to adapt to vips.

**Comment by lovell** [comment]
If you've not already seen it, this has also been discussed over at mozjpeg, upon which smallfry currently depends - see https://github.com/mozilla/mozjpeg/issues/12 and https://github.com/mozilla/mozjpeg/issues/110

**Comment by jcupitt** [comment]
Thank you Lovell, those are useful links.

Thinking about this more, I don't thinking building this into vips is the best idea. To do well it needs a deep knowledge of the encoder, which libvips does not have.

Since mozjpeg is thinking of adding this feature we should probably wait for their implementation. I'll leave this issue open as a reminder to add a control for this switch when the mozjpeg version is ready.

**Comment by rayshan** [comment]
Thanks guys, I'll keep an eye on the mozjpeg issues.
